File: Baekjoon/11053_가장 긴 증가하는 부분수열 길이/B_11053_4.py
def binary_search(start, end, target):
    if start >= end:
        return end

    mid = (start + end) // 2
    if lis[mid] < target:
        start = mid + 1
        return binary_search(start, end, target)
    else:
        if lis[mid] == target:
            return mid
        end = mid
        # end = mid - 1을 하면 안 된다.
        # start 가 0이고 end 가 1이었을때는 mid 가 0이라서 end 를 mid - 1로 하면 인덱스가 맞지 않게 된다.
        return binary_search(start, end, target)


N = int(input())
numbers = list(map(int, input().split()))

lis = [numbers[0]]
for i in range(1, len(numbers)):
    if numbers[i] > lis[-1]:
        lis.append(numbers[i])
    else:
        idx = binary_search(0, len(lis) - 1, numbers[i])
        lis[idx] = numbers[i]

print(len(lis))

# 참고
# https://chanhuiseok.github.io/posts/algo-49/
# https://shoark7.github.io/programming/algorithm/3-LIS-algorithms

File: Baekjoon/11053_가장 긴 증가하는 부분수열 길이/B_11053_4.js
const binarySearch = function(start, end, num) {
    const mid = Math.floor((start + end) / 2)

    // 탈출 조건이 핵심
    // if (start === mid) {
    // 위의 조건은 num이 가장 작을 경우를 고려하지 못한다
    if (start === end) {
        return end
    }

    if (dp[mid] > num) {
        return binarySearch(start, mid, num)
    } else {
        if (dp[mid] === num) {
            return mid
        }
        return binarySearch(mid + 1, end, num)
    }
}

const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const N = parseInt(input[0])
const sequence = input[1].split(' ').map(v => parseInt(v))

let dp = [sequence[0]]

for (let i = 1; i < N; i++) {
    const number = sequence[i]

    if (number > dp[dp.length - 1]) {
        dp.push(number)
    } else {
        const idx = binarySearch(0, dp.length - 1, number)
        dp[idx] = number
    }
}

console.log(dp.length)
